snuba/datasets/tags_column_processor.py

```diff
@@ -77,38 +125,135 @@ def __string_col(self, col: str) -> str:
+        )
+
+    def __extract_top_level_tag_conditions(self, condition: Expression) -> List[str]:
+        """
+        Finds the top level conditions that include tags_key in an expression.
+        For top level condition. In Snuba conditions are expressed in layers, the top level
+        ones are in AND, the nested ones are in OR and so on.
+        We can only apply the arrayFilter optimization to tag keys conditions that are not in
+        OR with other columns. To simplify the problem, we only consider those conditions that
+        are included in the first level of the query:
+        [['tagskey' '=' 'a'],['col' '=' 'b'],['col2' '=' 'c']]  works
+        [[['tagskey' '=' 'a'], ['col2' '=' 'b']], ['tagskey' '=' 'c']] does not
+        """
+
+        if is_binary_condition(condition, ConditionFunctions.EQ):
+            # This is to make mypy happy.
+            assert isinstance(condition, FunctionCall)
+            if (
+                isinstance(condition.parameters[0], Column)
+                and condition.parameters[0].column_name == "tags_key"
+                and isinstance(condition.parameters[1], Literal)
+            ):
+                # Tags are strings, but mypy does not know that
+                return [str(condition.parameters[1].value)]
+
+        if is_in_condition(condition):
+            assert isinstance(condition, FunctionCall)
+            if (
+                isinstance(condition.parameters[0], Column)
+                and condition.parameters[0].column_name == "tags_key"
+            ):
+                # The parameters of the inner function `a IN tuple(b,c,d)`
+                assert isinstance(condition.parameters[1], FunctionCall)
+                literals = condition.parameters[1].parameters
+                return [
+                    str(literal.value)
+                    for literal in literals
+                    if isinstance(literal, Literal)
+                ]
+
+        if is_binary_condition(condition, BooleanFunctions.AND):
+            assert isinstance(condition, FunctionCall)
+            return self.__extract_top_level_tag_conditions(
+                condition.parameters[0]
+            ) + self.__extract_top_level_tag_conditions(condition.parameters[1])
+
+        return []
+
+    def __get_filter_tags(self, query: Query) -> List[str]:
+        """
+        Identifies the tag names we can apply the arrayFilter optimization on.
+        Which means: if the tags_key column is in the select clause and there are
+        one or more top level conditions on the tags_key column.
+        """
+        if not state.get_config("ast_tag_processor_enabled", 0):
+            return []
+
+        select_clause = query.get_selected_columns_from_ast() or []
+
+        tags_key_found = any(
+            [
+                col.column_name == "tags_key"
+                for expression in select_clause
+                for col in expression
+                if isinstance(col, Column)
+            ]
```

Comment:
There is a similar comment in `snuba/query/conditions.py` (I didn't review these files in order) but this can be a generator instead of a list comprehension to allow an early exit if any of these evaluate as `True`.